lib/matplotlib/__init__.py

```diff
@@ -1515,6 +1517,261 @@ def test(verbosity=1):
+            if data is not None:
+                if arg_names_at_runtime:
+                    # update the information about replace names and
+                    # label position
+                    _arg_names = positional_parameter_names(args, data)
+                    if (label_namer  # we actually want a label here ...
+                        and _arg_names  # and we can find a label in *args ...
+                        and (label_namer in _arg_names)):  # and it is in *args
+                        _label_namer_pos = _arg_names.index(label_namer)
+                        if "label" in _arg_names:
+                            _label_pos = arg_names.index("label")
+
+                # save the current label_namer value so that it can be used as
+                # a label
+                if _label_namer_pos < len(args):
+                    label = args[_label_namer_pos]
+                else:
+                    label = kwargs.get(label_namer, None)
+                # ensure a string, as label can't be anything else
+                if not isinstance(label, six.string_types):
+                    label = None
+
+                if (replace_names is None) or (replace_all_args is True):
+                    # all should be replaced
+                    args = tuple(_replacer(data, a) for
+                                 j, a in enumerate(args))
+                else:
+                    # An arg is replaced if the arg_name of that position is
+                    #   in replace_names ...
+                    if len(_arg_names) < len(args):
+                        raise RuntimeError(
+                            "Got more args than function expects")
+                    args = tuple(_replacer(data, a)
+                                 if _arg_names[j] in replace_names else a
+                                 for j, a in enumerate(args))
+
+                if replace_names is None:
+                    # replace all kwargs ...
+                    kwargs = dict((k, _replacer(data, v))
+                                  for k, v in six.iteritems(kwargs))
+                else:
+                    # ... or only if a kwarg of that name is in replace_names
+                    kwargs = dict((k, _replacer(data, v)
+                                   if k in replace_names else v)
+                                  for k, v in six.iteritems(kwargs))
+
+            # replace the label if this func "wants" a label arg and the user
+            # didn't set one. Note: if the user puts in "label=None", it does
+            # *NOT* get replaced!
+            user_supplied_label = (
+                (len(args) >= _label_pos) or  # label is included in args
+                ('label' in kwargs)  # ... or in kwargs
+            )
+            if (label_namer and not user_supplied_label):
+                if _label_namer_pos < len(args):
+                    try:
+                        kwargs['label'] = args[_label_namer_pos].name
```

Comment:
There is a `get_label` function in cbook added as part of this PR, please use it here.  See my PR against your branch for the changes it needs.